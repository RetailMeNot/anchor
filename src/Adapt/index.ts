export { Adapt } from './Adapt.component';
