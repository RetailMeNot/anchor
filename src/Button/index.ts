export { Button } from './Button.component';
