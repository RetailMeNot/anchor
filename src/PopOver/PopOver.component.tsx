// VENDOR
import * as React from 'react';
import styled from '@xstyled/styled-components';
import classNames from 'classnames';
// COMPONENTS
import { Position } from '../utils/position/position';
import { PositionContainer } from '../utils/PositionContainer';
// UTILS
import { get } from '../utils/get/get';

interface PopOverContainerProps extends React.HTMLAttributes<HTMLDivElement> {
    arrowIndent?: string;
    arrowSize?: string;
    background?: string;
    border?: string;
    borderRadius?: string;
    color?: string;
    content: string | React.ReactChild;
    delay?: string;
    debug?: boolean;
    shadow?: string;
    spacing?: number;
    maxWidth?: string;
    position?: Position;
    showArrow?: boolean;
    active: boolean;
}

interface PopOverContainerState {
    height: number;
    width: number;
    containerHeight: number;
    containerWidth: number;
}

const StyledPopOver = styled('div')`
    position: relative;
`;

export class PopOver extends React.PureComponent<
    PopOverContainerProps,
    PopOverContainerState
> {
    public state = {
        width: 0,
        height: 0,
        containerHeight: 0,
        containerWidth: 0,
    };
    private readonly popOverContainerRef: React.RefObject<
        HTMLDivElement
    > = React.createRef();
    private readonly popOverRef: React.RefObject<
        HTMLDivElement
    > = React.createRef();
    // TODO: why the fuck not just use the code in the dropdown/abstract that out and inject it into two instances?
    componentDidMount(): void {
        const { current: popOver } = this.popOverRef;
        const { current: popOverContainer } = this.popOverContainerRef;
        this.setState({
            height: get(popOver, 'clientHeight', 0),
            width: get(popOver, 'clientWidth', 0),
            containerHeight: get(popOverContainer, 'clientHeight', 0),
            containerWidth: get(popOverContainer, 'clientWidth', 0),
        });
    }

    componentDidUpdate(prevProps: PopOverContainerProps) {
        const {
            showArrow: prevShowArrow,
            position: prevPosition,
            spacing: prevSpacing,
        } = prevProps;
        const { showArrow, position, spacing } = this.props;

        if (
            showArrow !== prevShowArrow ||
            position !== prevPosition ||
            spacing !== prevSpacing
        ) {
            const {
                current: dropDown,
            }: { current: any } = this.popOverRef;
            const {
                current: container,
            }: { current: any } = this.popOverContainerRef;

            if (container) {
                this.setState({
                    height: get(dropDown, 'clientHeight', 0),
                    width: get(dropDown, 'clientWidth', 0),
                    containerHeight: get(container, 'clientHeight', 0),
                    containerWidth: get(container, 'clientWidth', 0),
                });
            }
        }
    }

    render() {
        const {
            arrowIndent,
            arrowSize,
            background = '#000000',
            children,
            className,
            border,
            borderRadius,
            color = '#ffffff',
            content,
            delay,
            debug,
            shadow,
            spacing = 8,
            maxWidth,
            position = 'bottom',
            showArrow,
            active,
            ...props
        } = this.props;
        const { height, containerHeight, containerWidth, width } = this.state;
        return (
            <StyledPopOver
                className={classNames('anchor-pop-over', className)}
                {...props}
                ref={this.popOverRef}
            >
                {children}
                <PositionContainer
                    ref={this.popOverContainerRef}
                    active={active}
                    arrowIndent={arrowIndent}
                    arrowSize={arrowSize}
                    background={background}
                    border={border}
                    borderRadius={borderRadius}
                    children={content}
                    className="anchor-pop-over-element"
                    color={color}
                    containerHeight={containerHeight}
                    containerWidth={containerWidth}
                    height={height}
                    width={width}
                    delay={delay}
                    debug={debug}
                    shadow={shadow}
                    spacing={spacing}
                    maxWidth={maxWidth}
                    position={position}
                    showArrow={showArrow}
                />
            </StyledPopOver>
        );
    }
}
