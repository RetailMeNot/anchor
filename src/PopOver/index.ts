export { PopOver } from './PopOver.component';
