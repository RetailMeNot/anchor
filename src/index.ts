export { Avatar } from './Avatar';
export { AutoComplete } from './AutoComplete';
export { Button } from './Button/Button.component';
export { Badge } from './Badge';
export { DropDown } from './DropDown';
export { Container, Flex, Box } from './Grid';
export { Icon } from './Icon';
export { Page, Footer, Header } from './Layout';
export { List, Item as ListItem } from './List';
export { Menu, Item as MenuItem } from './Menu';
export { colors, fonts, sizes, variables, GlobalCSS, NormalizeCSS } from './theme';
