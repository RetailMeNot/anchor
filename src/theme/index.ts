export { default as colors } from './colors.theme';
export { default as fonts } from './fonts.theme';
export { default as sizes } from './sizes.theme';
export { default as variables } from './variables.theme';

export { NormalizeCSS, GlobalCSS } from './GlobalStyles';
