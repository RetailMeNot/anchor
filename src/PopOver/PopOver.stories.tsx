// REACT
import * as React from 'react';
// STORYBOOK
import { storiesOf } from '@storybook/react';
// VENDOR
import styled, { ThemeProvider } from '@xstyled/styled-components';
// ANCHOR
import { RootTheme } from '../../src/theme';
import { PopOver } from './PopOver.component';

const StyledStory = styled('div')`
    padding: 2rem;
    color: text.base;
    font-family: base;
`;

storiesOf('Components/PopOver', module).add('Default', () => (
    <ThemeProvider theme={RootTheme}>
        <StyledStory>
            <PopOver />
        </StyledStory>
    </ThemeProvider>
));
