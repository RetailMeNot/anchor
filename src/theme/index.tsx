// Theme Files
import { typography } from './typography.theme';
import { ColorsTheme } from './colors.theme';
import { ThemeProvider as XstyledThemeProvider } from '@xstyled/styled-components';
import styled from '@xstyled/styled-components';
import { space as spaceStyles, SpaceProps } from '@xstyled/system';
import { BUTTON_KEY, BUTTON_THEME } from '../Button';
import { INPUT_KEY, INPUT_THEME } from '../Form/Input/utils';
import React from 'react';
import {Button} from '../Button';
import classNames from 'classnames';

export const RootTheme = {
    typography,
    radii: { none: '0', base: '4px', modal: '8px', circular: '1000px' },
    [BUTTON_KEY]: BUTTON_THEME,
    [INPUT_KEY]: INPUT_THEME,
    colors: ColorsTheme,
    fonts: {
        base:
            'Avenir Next, Segoe UI, Helvetica Neue, Helvetica, Roboto, sans-serif',
        mono: `'SFMono-Regular', Consolas, Liberation Mono, Menlo, Courier, monospace`,
    },
    // These are all minimums
    breakpoints: {
        xs: 0,
        sm: 576,
        md: 768,
        lg: 992,
        xl: 1200,
        xxl: 1600,
    },
    // todo: we should likely remove the color.borders colors and just use this
    // or find a way to reference the colors.border colors here
    // e.g. `thin solid borders.base`
    borders: {
        light: `thin solid ${ColorsTheme.borders.light}`,
        base: `thin solid ${ColorsTheme.borders.base}`,
        dark: `thin solid ${ColorsTheme.borders.dark}`,
    },

    skeleton: {
        colorStart: '#E7E7E7',
        colorEnd: '#D3D3D3',
    },
};

// Exporting out @xstyled's ThemeProvider. Only way I could get it to work w/o issues.
// export class ThemeProvider2 extends XstyledThemeProvider {}

interface IterableObject {
    [s: number]: number | string | boolean | IterableObject;
  }

interface DebugProps {
    property: number | string | boolean | IterableObject;
    propertyName: string;
    rootProperty?: boolean;
}

interface StyledDebugProps {
    indent?: number;
}

const StyledDebug = styled('div')<StyledDebugProps>`
    font-family: mono;
    margin-top: 0.5rem;
    color: white;

    .category .category h2 {
        background-color: #ec8947;
    }

    .category .category .box {
        border-color: #ec8947;
    }

    .category .category .category h2 {
        background-color: #b84343;
        color: white;
    }

    .category .category .category .box {
        border-color: #b84343;
    }

    .category .category .category .category h2 {
        background-color: #91cdd2;
        color: black;
    }

    .category .category .category .category .box {
        border-color: #91cdd2;
    }

    .category .category .category .category .category h2 {
        background-color: #567224;
        color: white;
    }

    .category .category .category .category .category .box {
        border-color: #567224;
    }
`;

const StyledH2 = styled('h2')`
    ${spaceStyles}
    background-color: #ffea92;
    color: black;
    margin-bottom: 0;
    margin-top: 0;
    padding: 0.25rem 0.5rem 0.25rem 1rem;
    font-size: 1rem;
    cursor: pointer;
    border-radius: 0 1rem 1rem 0;
`;

interface BoxProps extends SpaceProps {
    show?: boolean;
}

const Box = styled('div')<BoxProps>`
    ${spaceStyles}
    width: 20rem;
    border-left: solid #ffea92 2px;
    margin: 0.5rem;
    ${({show}) => `display: ${show ? 'block' : 'none'}`}
`;

const Category = styled('div')<SpaceProps>`
    ${spaceStyles}
    width: 20rem;
`;


const ColorBox = styled('div')<any>`
    width: 1rem;
    height: 1rem;
    display: inline-block;
    vertical-align: middle;
    margin-left: 0.5rem;
    background-color: ${({backgroundColor}) => backgroundColor};
    border: solid white 1px;
`;

const Debug = ({property, propertyName}: DebugProps) => {
    const [show, setShow] = React.useState(false);

    return(
        <StyledDebug>
            {(property || property === 0)
                ? (
                    typeof property === 'number' ||
                    typeof property === 'string' ||
                    typeof property === 'boolean' ||
                    typeof property === 'function'
                        ? (
                                <>
                                    <span style={{fontWeight: 'bold'}}>{propertyName}: </span>
                                    <span>
                                        {typeof property === 'function'
                                            ? 'function'
                                            : /^#[a-zA-Z0-9]{4,7}/.test(property.toString())
                                                ? (<>
                                                    <span style={{fontStyle: 'italic'}}>{property.toString()}</span>
                                                    <ColorBox backgroundColor={property} />
                                                    </>)
                                                : <span style={{fontStyle: 'italic'}}>{property.toString()}</span>
                                        }
                                    </span>
                                </>
                            )
                        :
                            <Category pl={0} className="category">
                                <StyledH2 onClick={() => setShow(!show)}>{propertyName}</StyledH2>
                                <Box pl={2} className="box" show={show}>
                                    {Object.values(property).map((_property, index) => (
                                        <Debug
                                            key={index}
                                            property={_property}
                                            propertyName={Object.getOwnPropertyNames(property)[index]}
                                        />
                                    ))}
                                </Box>
                            </Category>
                    )
                : (
                    <>
                       Empty.
                    </>
                )
            }
        </StyledDebug>
    );
};

const DebugWrapper = styled('div')`
    background-color: rgba(0,0,0,.75);
    border-left-color: transparent;
    position: fixed;
    left: 0;
    top: 0;
    height: 100%;
    overflow: hidden;
    width: 30rem;
    padding: 1rem 2rem 1rem 0;

    &.closed {
        height:2.4rem;
        width: auto;
        padding-right: 1rem;

        .debug-scroll {
            display: none;
        }
    }
`;

const DebugScroll = styled('div')`
    height: 100%;
    overflow-y: auto;
    width: 100%;
`;

interface ThemeProviderProps {
    children?: any;
    debug?: boolean;
    theme?: any;
}

export const ThemeProvider = ({ children, debug, theme }: ThemeProviderProps) => {
    const isOpen = localStorage.themeProviderDebug === 'true';
    const [isOpenState, setIsOpenState] = React.useState(isOpen);

    return (
        <XstyledThemeProvider theme={theme}>
            {debug &&
                <DebugWrapper className={classNames(!isOpenState ? 'closed' : null)}>
                    <Button
                        onClick={() => {
                            localStorage.themeProviderDebug = !isOpenState;
                            setIsOpenState(!isOpenState);
                        }}
                        size="sm"
                    >
                        ThemeProvider Debug
                    </Button>

                    <DebugScroll className="debug-scroll">
                        <Debug property={theme} propertyName="Theme" />
                    </DebugScroll>
                </DebugWrapper>
            }
            {children}
        </XstyledThemeProvider>
    );
};
