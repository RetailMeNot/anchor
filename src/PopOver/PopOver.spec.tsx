// VENDOR
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { ThemeProvider } from '@xstyled/styled-components';
import { shallow, mount } from 'enzyme';
// // ANCHOR
import { RootTheme } from '../../src/theme';
import { PopOver } from './PopOver.component';
import { Button } from '../Button';

describe('Component: PopOver', () => {
    it('should be defined.', () => {
        const subject = (
            <ThemeProvider theme={RootTheme}>
                <PopOver active={true} position="bottomStart" content="Hello!">
                    <Button />
                </PopOver>
            </ThemeProvider>
        );
        const wrapper = mount(subject);
        const component = shallow(subject);

        expect(subject).toBeDefined();
        expect(wrapper).toBeDefined();
        expect(component).toBeDefined();
        const tree = renderer.create(subject).toJSON();
        expect(tree).toMatchSnapshot();
        expect(true);
    });
});
