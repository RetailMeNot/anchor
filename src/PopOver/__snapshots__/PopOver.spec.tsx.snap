// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: PopOver should be defined. 1`] = `
<div
  className="anchor-popover"
/>
`;
